futures-util/src/task/local_waker_ref.rs

```diff
@@ -0,0 +1,140 @@
+#![allow(cast_ptr_alignment)] // clippy is too strict here
+
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::ptr::NonNull;
+use std::sync::Arc;
+use std::task::{LocalWaker, Waker, Wake, UnsafeWake};
+
+/// A [`LocalWaker`](::std::task::LocalWaker) that is only valid for a given lifetime.
+///
+/// Note: this type implements [`Deref<Target = LocalWaker>`](::std::ops::Deref),
+/// so it can be used to get a `&LocalWaker`.
+#[derive(Debug)]
+pub struct LocalWakerRef<'a> {
+    local_waker: LocalWaker,
+    _marker: PhantomData<&'a ()>,
+}
+
+impl<'a> LocalWakerRef<'a> {
+    /// Create a new [`LocalWakerRef`] from a [`LocalWaker`].
+    ///
+    /// Note: this function is safe, but it is generally only used
+    /// from `unsafe` contexts that need to create a `LocalWaker`
+    /// that is guaranteed not to outlive a particular lifetime.
+    pub fn new(local_waker: LocalWaker) -> Self {
+        LocalWakerRef {
+            local_waker,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<'a> Deref for LocalWakerRef<'a> {
+    type Target = LocalWaker;
+
+    fn deref(&self) -> &LocalWaker {
+        &self.local_waker
+    }
+}
+
+// Pointers to this type below are really pointers to `Arc<W>`
+struct ReferencedArc<W> {
+    _marker: PhantomData<W>,
+}
+
+unsafe impl<W: Wake + 'static> UnsafeWake for ReferencedArc<W> {
+    #[inline]
+    unsafe fn clone_raw(&self) -> Waker {
+        let me = self as *const ReferencedArc<W> as *const Arc<W>;
+        Arc::clone(&*me).into()
+    }
+
+    #[inline]
+    unsafe fn drop_raw(&self) {}
+
+    #[inline]
+    unsafe fn wake(&self) {
+        let me = self as *const ReferencedArc<W> as *const Arc<W>;
+        W::wake(&*me)
```

Comment:
here

Reply:
Hm, I guess this is arguable, but I actually like using UFCS here and below because (1) I think it makes it clearer which implementation of the `Wake` trait is being delegated to and (2) it makes it clear that an unsafe re-referencing of a raw pointer is occuring. In any case, it really doesn't matter that much, so I guess I'm fine changing it?

Reply:
I'd use the method call syntax because it's easier to read (IMO). It's however fine by me if you leave it like that. If you decide to keep it, consider changing the ones in `futures_unordered/task.rs` as well then for consistency.